network/src/peer_store/db.rs

```diff
@@ -0,0 +1,313 @@
+            peer_id: PeerId::from_bytes(row.get(1)).expect("parse peer_id"),
+            connected_addr: Multiaddr::from_bytes(row.get(2)).expect("parse multiaddr"),
+            score: row.get(3),
+            status: Status::from(row.get::<_, String>(4)),
+            endpoint: bool_to_endpoint(row.get::<_, bool>(5)),
+            ban_time: secs_to_duration(row.get(6)),
+            connected_time: secs_to_duration(row.get(7)),
+        })
+        .expect("query");
+    rows.next().map(|row| row.expect("query first"))
+}
+
+pub fn contains_most_peers_network_group(conn: &Connection) -> Vec<PeerInfo> {
+    let (network_group, group_peers_count) = conn.query_row::<(Vec<u8>, u32), _, _>("SELECT network_group, COUNT(network_group) AS network_group_count FROM peer_info GROUP BY network_group ORDER BY network_group_count DESC LIMIT 1", NO_PARAMS, |r| (r.get(0), r.get(1))).expect("query count");
+    let mut stmt = conn.prepare("SELECT id, peer_id, connected_addr, score, status, endpoint, ban_time, connected_time FROM peer_info WHERE network_group=:network_group").expect("prepare");
+    let rows = stmt
+        .query_map_named(&[(":network_group", &network_group)], |row| PeerInfo {
+            id: row.get(0),
+            peer_id: PeerId::from_bytes(row.get(1)).expect("parse peer_id"),
+            connected_addr: Multiaddr::from_bytes(row.get(2)).expect("parse multiaddr"),
+            score: row.get(3),
+            status: Status::from(row.get::<_, String>(4)),
+            endpoint: bool_to_endpoint(row.get::<_, bool>(5)),
+            ban_time: secs_to_duration(row.get(6)),
+            connected_time: secs_to_duration(row.get(7)),
+        })
+        .expect("query");
+    let mut peers = Vec::with_capacity(group_peers_count as usize);
+    for row in rows {
+        peers.push(row.expect("extra value from query result"));
+    }
+    peers
+}
+
+pub fn peer_info_count(conn: &Connection) -> u32 {
+    conn.query_row::<u32, _, _>("SELECT COUNT(*) FROM peer_info", NO_PARAMS, |r| r.get(0))
+        .expect("query count")
+}
+
+pub fn get_peer_addrs(conn: &Connection, id: u32, count: u32) -> Vec<Multiaddr> {
+    let mut stmt = conn
+        .prepare("SELECT addr FROM peer_addr WHERE peer_info_id == :id LIMIT :count")
+        .expect("prepare");
+    let rows = stmt
+        .query_map_named(&[(":id", &id), (":count", &count)], |row| {
+            Multiaddr::from_bytes(row.get(0)).expect("parse multiaddr")
+        })
+        .expect("query");
+    let mut addrs = Vec::with_capacity(count as usize);
+    for row in rows {
+        addrs.push(row.expect("extra value from query result"));
+    }
+    addrs
+}
+
+pub fn get_peers_to_attempt(conn: &Connection, count: u32) -> Vec<(PeerId, Multiaddr)> {
+    // random select peers
+    let mut stmt = conn.prepare("SELECT id, peer_id FROM peer_info WHERE status != :connected_status AND ban_time < strftime('%s','now') ORDER BY RANDOM() LIMIT :count").expect("prepare");
```

Comment:
`order by rand()` will do full table scan, need to test performance with huge table.

Reply:
Check the benchmark result in my comment.